clang/runtime/dpct-rt/include/dpct/math.hpp

```diff
@@ -242,6 +243,66 @@ inline unsigned unordered_compare_mask(const sycl::marray<T, 2> a,
       .as<sycl::vec<unsigned, 1>>();
 }
 
+/// Bitfield-extract.
+///
+/// Extract bit field from \param source and return the zero or sign-extended
+/// result. Source \param bit_start gives the bit field starting bit position,
+/// and source \param num_bits gives the bit field length in bits.
+///
+/// The result is padded with the sign bit of the extracted field. If the start
+/// position is beyond the msb of the input, the result was filled with the
+/// replicated sign bit of the extracted field.
+///
+/// \tparam T The type of \param source value, must be an integer.
+/// \param source The source value to extracting.
+/// \param bit_start The position to start extracting.
+/// \param num_bits The number of bits to extracting.
+template <typename T>
+inline std::enable_if_t<std::is_integral_v<T>, T>
+bfe(T source, uint32_t bit_start, uint32_t num_bits) {
+  unsigned max_bits_num = CHAR_BIT * sizeof(T);
+  unsigned pos = std::min(bit_start, max_bits_num);
+  unsigned len = std::min(pos + num_bits, max_bits_num) - pos;
+  T mask = (T{1} << len) - 1;
+  if constexpr (std::is_signed_v<T>) {
+    // Find the sign-bit, the result is padded with the sign bit of the
+    // extracted field.
+    //
+    // sign_bit = len == 0 ? 0 : source[min(pos + len - 1, max_bits_num - 1)]
+    unsigned sign_bit_pos = std::min((uint8_t)bit_start + (uint8_t)num_bits - 1,
+                                     (uint8_t)max_bits_num - 1);
+    T sign_bit = len != 0 && ((source >> sign_bit_pos) & 1);
+    return ((source >> pos) & mask) | (-sign_bit & ~mask);
+    //                              ^^^^^^^^^^^^^^^^^^^^^
+    //                              pad result with sign-bit.
+  } else {
+    // Is T is an unsigned integer type, the sign-bit does not need to be
+    // padded.
+    return (source >> pos) & mask;
+  }
+}
+
+/// Bitfield-insert.
+///
+/// Align and insert a bit field from \param x into \param y . Source \param
+/// bit_start gives the starting bit position for the insertion, and source
+/// \param num_bits gives the bit field length in bits.
+///
+/// \tparam T The type of \param x and \param y , must be an unsigned integer.
+/// \param x The source of the bitfield.
+/// \param y The source where bitfield is inserted.
+/// \param bit_start The position to start insertion.
+/// \param num_bits The number of bits to insertion.
+template <typename T>
+inline std::enable_if_t<std::is_unsigned_v<T>, T>
+bfi(T x, T y, uint32_t bit_start, uint32_t num_bits) {
```

Comment:
please add const for the args which will not changed.

Reply:
done